src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs

```diff
@@ -35,15 +35,58 @@ public static partial class Base64
         /// - InvalidData - if the input contains bytes outside of the expected base64 range, or if it contains invalid/more than two padding characters,
         ///   or if the input is incomplete (i.e. not a multiple of 4) and <paramref name="isFinalBlock"/> is <see langword="true"/>.
         /// </returns>
-        public static unsafe OperationStatus DecodeFromUtf8(ReadOnlySpan<byte> utf8, Span<byte> bytes, out int bytesConsumed, out int bytesWritten, bool isFinalBlock = true)
+        public static OperationStatus DecodeFromUtf8(ReadOnlySpan<byte> utf8, Span<byte> bytes, out int bytesConsumed, out int bytesWritten, bool isFinalBlock = true)
         {
-            if (utf8.IsEmpty)
+            OperationStatus status = OperationStatus.Done;
+            bytesConsumed = 0;
+            bytesWritten = 0;
+
+            while (!utf8.IsEmpty)
             {
-                bytesConsumed = 0;
-                bytesWritten = 0;
-                return OperationStatus.Done;
+                status = DecodeFromUtf8Core(utf8, bytes, out int localConsumed, out int localWritten, isFinalBlock);
+                bytesConsumed += localConsumed;
+                bytesWritten += localWritten;
+
+                if (status is not OperationStatus.InvalidData or OperationStatus.DestinationTooSmall)
+                {
+                    break;
+                }
+
+                utf8 = utf8.Slice(localConsumed);
+                bytes = bytes.Slice(localWritten);
+
+                if (!TrySkipWhiteSpace(utf8, out localConsumed))
+                {
+                    if (localConsumed > 0)
+                    {
+                        bytesConsumed += localConsumed;
+                        status = OperationStatus.Done;
+                    }
+
+                    // The end of the input has been reached.
+                    break;
+                }
+
+                if (localConsumed == 0)
+                {
+                    // First char isn't whitespace, but we didn't consume anything,
+                    // thus the input may have whitespace anywhere in between. So fall back to block-wise decoding.
+                    return DecodeWithWhiteSpaceBlockwise(utf8, bytes, out bytesConsumed, out bytesWritten, isFinalBlock);
```

Comment:
This looks wrong as it's throwing away the number of consumed/written bytes in previous iterations.
I'm guessing input like `"MTIzN DU2"` would hit this (valid block + whitespace in the middle of a block).

Reply:
It does _look_ wrong indeed, but It turns out it's correct but for bad / very unobvious reasons.  I hadn't looked at this part of the code before, but it was written such that DecodeWithWhiteSpaceBlockwise used Unsafe.SkipInit on the out arguments and then continued to add to them; as a result, the original bytesConsumed and bytesWritten were serving as the seed values.  I'll fix it to use ref instead of out.